File: typescript/medium/19.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

// @ts-ignore
function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  // @ts-ignore
  let dummyNode = new ListNode();
  dummyNode.next = head;
  let left = dummyNode;
  let right = head;

  while (n > 0 && right) {
    right = right.next;
    n--;
  }

  while (right) {
    right = right.next;
    left = left.next;
  }

  left.next = left.next.next;

  return dummyNode.next;
}

File: python/19.py
# 19. Remove Nth Node From End of List

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        p1 = head
        p2 = head

        for i in range(n):
            p1 = p1.next

        if p1 is None:
            return head.next

        while p1.next is not None:
            p1 = p1.next
            p2 = p2.next

        p2.next = p2.next.next

        return head